hibernate-reactive-core/src/test/java/org/hibernate/reactive/CompletionStagesTest.java

```diff
@@ -0,0 +1,177 @@
+
+import org.assertj.core.api.Assertions;
+
+public class CompletionStagesTest {
+
+	static int loopCount = 0;
+
+	@Test
+	public void testTotalWithIntegers() {
+		// total(int start, int end, Function<Integer,CompletionStage<Integer>> consumer)
+		int startInt = 0;
+		int endInt = 11;
+
+		CompletionStages.total(
+				startInt, endInt,
+				i -> CompletionStages.completedFuture( 1 )
+		).thenAccept( total -> Assertions.assertThat( total ).isEqualTo( 11 ) );
+
+	}
+
+	@Test
+	public void testTotalWithIterator() {
+		// total(Iterator<T> iterator, Function<T,CompletionStage<Integer>> consumer)
+		List<Object> entries = new ArrayList<>();
+		entries.add("a");
+		entries.add("b");
+		entries.add("c");
+		entries.add("d");
+		entries.add("e");
+
+		CompletionStages.total(
+				entries.iterator(),
+				i -> CompletionStages.completedFuture( 1 )
+		).thenAccept( total -> Assertions.assertThat( total ).isEqualTo( entries.size() ) );
+	}
+
+	@Test
+	public void testTotalWithArray() {
+		// total(T[] array, Function<T,CompletionStage<Integer>> consumer)
+		String[] entries = { "a", "b", "c", "d", "e" };
+
+		CompletionStages.total(
+				entries,
+				i -> CompletionStages.completedFuture( 1 )
+		).thenAccept( total -> Assertions.assertThat( total ).isEqualTo( entries.length ) );
+	}
+
+	@Test
+	public void testLoopOnArray() {
+		// loop(T[] array, Function<T,CompletionStage<?>> consumer)
+		String[] entries = { "a", "b", "c", "d", "e" };
+		loopCount = 0;
+		CompletionStages.loop(
+				entries,
+				index -> {
+					loopCount++;
+					return CompletionStages.voidFuture();
+				}
+		).thenAccept( count -> Assertions.assertThat( loopCount ).isEqualTo( entries.length ) );
+	}
+
+	@Test
+	public void testLoopOnIterator() {
+		// loop(Iterator<T> iterator, Function<T,CompletionStage<?>> consumer)
+		List<Object> entries = new ArrayList<>();
+		entries.add("a");
+		entries.add("b");
+		entries.add("c");
+		entries.add("d");
+		entries.add("e");
+		loopCount = 0;
+		CompletionStages.loop(
+				entries.iterator(),
+				index -> {
+					loopCount++;
+					return CompletionStages.voidFuture();
+				}
+		).thenAccept( count -> Assertions.assertThat( loopCount ).isEqualTo( entries.size() ) );
+	}
+
+	@Test
+	public void testLoopOnIterable() {
+		// loop(Iterable<T> iterable, Function<T,CompletionStage<?>> consumer)
+		List<Object> entries = new ArrayList<>();
+		entries.add("a");
+		entries.add("b");
+		entries.add("c");
+		entries.add("d");
+		entries.add("e");
+		loopCount = 0;
+		CompletionStages.loop(
+				entries,
+				index -> {
+					loopCount++;
+					return CompletionStages.voidFuture();
+				}
+		).thenAccept( count -> Assertions.assertThat( loopCount ).isEqualTo( entries.size() ) );
+	}
+
+	@Test
+	public void testLoopOnStream() {
+		// loop(Stream<T> stream, Function<T,CompletionStage<?>> consumer)
+		String[] entries = { "a", "b", "c", "d", "e" };
+		loopCount = 0;
+		CompletionStages.loop(
+				Arrays.stream(entries),
+				index -> {
```

Comment:
This is misleading. This is not an index, this is an entry. You could rewrite almost all the tests to something like this:
```

	@Test
	public void testLoopOnStream() {
		String[] entries = { "a", "b", "c", "d", "e" };
		List<String> looped = new ArrayList<>();

		loop( Arrays.stream( entries ), entry -> completedFuture( looped.add( entry ) ) )
				.thenAccept( count -> assertThat( looped ).containsExactly( entries ) );
	}
```

Reply:
ah.. right.